From 4fae2309ff574d452aa3b2967b06f49b0d7a81d8 Mon Sep 17 00:00:00 2001
From: "bo.yang" <bo.yang@amlogic.com>
Date: Fri, 28 Mar 2014 11:23:24 +0800
Subject: [PATCH 3828/5965] encoder: prefect encoder function

by jiyu.yang@amlogic.com & wanhu.zheng@amlogic.com
supported by brian.zhu@amlogic.com
---
 arch/arm/boot/dts/amlogic/meson6tvd_ref.dtd   |   2 +-
 drivers/amlogic/amports/encoder.c             |  56 +--
 drivers/amlogic/amports/encoder.h             |  12 +-
 .../m6tvd/ucode/encoder/full_encoder_linux.h  | 347 +++++++++---------
 .../m6tvd/ucode/encoder/half_encoder_linux.h  | 270 +++++++-------
 drivers/amlogic/amports/vdec_reg.h            |   4 +-
 include/linux/amlogic/amports/canvas.h        |   6 +-
 7 files changed, 327 insertions(+), 370 deletions(-)

diff --git a/arch/arm/boot/dts/amlogic/meson6tvd_ref.dtd b/arch/arm/boot/dts/amlogic/meson6tvd_ref.dtd
index 0eea23bf41a9..7cb68c338e36 100755
--- a/arch/arm/boot/dts/amlogic/meson6tvd_ref.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6tvd_ref.dtd
@@ -194,7 +194,7 @@
 //$$ MODULE = "DISP&MM-AVC Video Encoder"
 //$$ DEVICE="amvenc_avc"
 //$$ L2 PROP_STR = "status"
-//$$ L3 PROP_U32 2 ="reg"
+//$$ L3 PROP_U32 ="reserve-memory"
 	amvenc_avc {
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
diff --git a/drivers/amlogic/amports/encoder.c b/drivers/amlogic/amports/encoder.c
index 00fc4c31d9d7..5aa8eef21cd9 100755
--- a/drivers/amlogic/amports/encoder.c
+++ b/drivers/amlogic/amports/encoder.c
@@ -114,14 +114,12 @@ static u32 log2_max_frame_num =4 ;
 static u32 anc0_buffer_id =0;
 static u32 qppicture  =26;
 static u32 process_irq = 0;
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 static u32 ie_me_mb_type  = 0;
 static u32 ie_me_mode  = 0;
 static u32 me_start_position = 0;
 static u32 ie_pippeline_block = 3;
 static u32 ie_cur_ref_sel = 0;
 static u32 half_ucode_mode = 0;
-#endif
 static int encode_inited = 0;
 static int encode_opened = 0;
 static int encoder_status = 0;
@@ -415,7 +413,6 @@ static void avc_canvas_init(void)
     debug_level(0,"dblk_buf_canvas is %d ; ref_buf_canvas is %d \n",dblk_buf_canvas , ref_buf_canvas);
 }
 
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 static void avc_init_ie_me_parameter(void)
 {
     me_start_position = 0;
@@ -620,7 +617,6 @@ static int  set_input_format (amvenc_mem_type type, amvenc_frame_fmt fmt, unsign
         mfdin_basic(input,iformat,oformat,picsize_x,picsize_y,r2y_en);
     return ret;
 }
-#endif
 
 static void encode_isr_tasklet(ulong data)
 {
@@ -642,7 +638,7 @@ static void encode_isr_tasklet(ulong data)
 static irqreturn_t enc_isr(int irq, void *dev_id)
 {
 
-	WRITE_HREG(HCODEC_ASSIST_MBOX2_CLR_REG, 1);
+	WRITE_HREG(HCODEC_IRQ_MBOX_CLR, 1);
 	encoder_status  = READ_HREG(ENCODER_STATUS);
 	if((encoder_status == ENCODER_IDR_DONE)
 	||(encoder_status == ENCODER_NON_IDR_DONE)
@@ -819,7 +815,7 @@ static void avc_prot_init(void)
     WRITE_HREG(QDCT_MB_CONTROL,
                 (1<<9) | // mb_info_soft_reset
                 (1<<0)); // mb read buffer soft reset
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8              
+
     if(half_ucode_mode == 0){
         WRITE_HREG(QDCT_MB_CONTROL,
                   (0<<28) | // ignore_t_p8x8
@@ -861,28 +857,7 @@ static void avc_prot_init(void)
                   (1<<1) | // mb_read_en
                   (0<<0));   // soft reset
     }
-#else
-    WRITE_HREG(QDCT_MB_CONTROL,
-              (0<<28) | // ignore_t_p8x8
-              (0<<27) | // zero_mc_out_null_non_skipped_mb
-              (0<<26) | // no_mc_out_null_non_skipped_mb
-              (0<<25) | // mc_out_even_skipped_mb
-              (0<<24) | // mc_out_wait_cbp_ready
-              (0<<23) | // mc_out_wait_mb_type_ready
-              (1<<22) | // i_pred_int_enable
-              (1<<19) | // i_pred_enable
-              (1<<20) | // ie_sub_enable
-              (1<<18) | // iq_enable
-              (1<<17) | // idct_enable
-              (1<<14) | // mb_pause_enable
-              (1<<13) | // q_enable
-              (1<<12) | // dct_enable
-              (1<<10) | // mb_info_en
-              (avc_endian<<3) | // endian
-              (1<<1) | // mb_read_en
-              (0<<0));   // soft reset
-#endif
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
+
     if(half_ucode_mode == 0){
         WRITE_HREG(SAD_CONTROL,
                   (0<<3) | // ie_result_buff_enable
@@ -944,7 +919,6 @@ static void avc_prot_init(void)
 
         WRITE_HREG(IE_DATA_FEED_BUFF_INFO,0);
     }
-#endif
 
     WRITE_HREG(HCODEC_CURR_CANVAS_CTRL,0);
     //debug_level(0,"current endian is %d \n" , avc_endian);
@@ -953,10 +927,10 @@ static void avc_prot_init(void)
     WRITE_HREG(VLC_CONFIG , data32);	
     
     /* clear mailbox interrupt */
-    WRITE_HREG(HCODEC_ASSIST_MBOX2_CLR_REG, 1);
+    WRITE_HREG(HCODEC_IRQ_MBOX_CLR, 1);
 
     /* enable mailbox interrupt */
-    WRITE_HREG(HCODEC_ASSIST_MBOX2_MASK, 1);
+    WRITE_HREG(HCODEC_IRQ_MBOX_MASK, 1);
 }
 
 void amvenc_reset(void)
@@ -1118,14 +1092,14 @@ bool amvenc_avc_on(void)
 
 #endif
 
-#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON6TV
+#if MESON_CPU_TYPE < MESON_CPU_TYPE_MESON8
 #define  DMC_SEC_PORT8_RANGE0  0x840
 #define  DMC_SEC_CTRL  0x829
 #endif
 
 void enable_hcoder_ddr_access(void)
 {
-#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON6TV
+#if MESON_CPU_TYPE < MESON_CPU_TYPE_MESON8
 	WRITE_SEC_REG(DMC_SEC_PORT8_RANGE0 , 0xffff);
 	WRITE_SEC_REG(DMC_SEC_CTRL , 0x80000000);
 #endif
@@ -1216,6 +1190,8 @@ static s32 avc_poweroff(void)
 	switch_mod_gate_by_name("vdec", 0);
 #else
 	hvdec_clock_disable();
+	switch_mod_gate_by_name("vdec", 0);
+
 #endif
 	return 0;
 }
@@ -1228,10 +1204,8 @@ static s32 avc_init(void)
     avc_canvas_init();
     WRITE_HREG(HCODEC_ASSIST_MMC_CTRL1,0x2);
     debug_level(1,"start to load microcode\n");
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
     if(half_ucode_mode == 1)
         p = half_encoder_mc;
-#endif
     if (amvenc_loadmc(p) < 0) {
         //amvdec_disable();
         return -EBUSY;
@@ -1253,10 +1227,8 @@ static s32 avc_init(void)
     avc_init_dblk_buffer(dblk_buf_canvas);   //decoder buffer , need set before each frame start
     avc_init_reference_buffer(ref_buf_canvas); //reference  buffer , need set before each frame start
     avc_init_assit_buffer(); //assitant buffer for microcode
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
     if(half_ucode_mode == 0)
         avc_init_ie_me_parameter();
-#endif
     WRITE_HREG(ENCODER_STATUS , ENCODER_IDLE);
     amvenc_start();
     encode_inited = 1;
@@ -1298,14 +1270,12 @@ void amvenc_avc_start_cmd(int cmd, unsigned* input_info)
 	}
 	avc_init_dblk_buffer(dblk_buf_canvas);   
 	avc_init_reference_buffer(ref_buf_canvas); 
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 	if(half_ucode_mode == 0){
 		if((cmd == ENCODER_IDR)||(cmd == ENCODER_NON_IDR)){
 			set_input_format((amvenc_mem_type)input_info[0], (amvenc_frame_fmt)input_info[1], input_info[2], input_info[3], input_info[4],(unsigned char)input_info[5]);
 		}
 		avc_init_ie_me_parameter();
 	}
-#endif
 	encoder_status = cmd;
 	WRITE_HREG(ENCODER_STATUS , cmd);
 	if((cmd == ENCODER_IDR)||(cmd == ENCODER_NON_IDR)){
@@ -1437,16 +1407,12 @@ static long amvenc_avc_ioctl(struct file *file,
 		break;    
 	case AMVENC_AVC_IOC_NEW_CMD:
 		amrisc_cmd = *((unsigned*)arg) ;
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 		if(half_ucode_mode == 0){
 			addr_info = (unsigned*)arg;
 			amvenc_avc_start_cmd(amrisc_cmd, &addr_info[1]);
 		}else{
 			amvenc_avc_start_cmd(amrisc_cmd, NULL);
 		}
-#else
-		amvenc_avc_start_cmd(amrisc_cmd, NULL);
-#endif
 		break;
 	case AMVENC_AVC_IOC_GET_STAGE:
 		*((unsigned*)arg)  = encoder_status;
@@ -1470,14 +1436,12 @@ static long amvenc_avc_ioctl(struct file *file,
 		    encoder_height = *((unsigned*)arg) ;
 		break;	
 	case AMVENC_AVC_IOC_CONFIG_INIT:
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 		addr_info = (unsigned*)arg;
 		if(*addr_info == 1)
 			half_ucode_mode = 1;
 		else
 			half_ucode_mode = 0;
 		debug_level(1,"avc init as mode %d\n",half_ucode_mode);
-#endif
 		avc_init();
 		break;		
 	case AMVENC_AVC_IOC_FLUSH_CACHE:
@@ -1538,14 +1502,12 @@ static long amvenc_avc_ioctl(struct file *file,
 		addr_info[11] = gAmvencbuff.bufspec->bitstream.buf_start;
 		addr_info[12] = gAmvencbuff.bufspec->bitstream.buf_size;
 		break;
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 	case AMVENC_AVC_IOC_SET_IE_ME_MB_TYPE:
 		ie_me_mb_type = *((unsigned*)arg);
 		break;
 	case AMVENC_AVC_IOC_SET_ME_PIXEL_MODE:
 		ie_me_mode |= (*((unsigned*)arg) & ME_PIXEL_MODE_MASK)<<ME_PIXEL_MODE_SHIFT;
 		break;
-#endif
 	case AMVENC_AVC_IOC_GET_DEVINFO:
 		strncpy((char *)arg,AMVENC_DEV_VERSION,strlen(AMVENC_DEV_VERSION));
 		break;
diff --git a/drivers/amlogic/amports/encoder.h b/drivers/amlogic/amports/encoder.h
index 1feb9c57b32a..eb47291fa2cf 100755
--- a/drivers/amlogic/amports/encoder.h
+++ b/drivers/amlogic/amports/encoder.h
@@ -2,12 +2,16 @@
 #ifndef __H264_H__
 #define __H264_H__
 
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 #define AMVENC_DEV_VERSION "AML-M8"
+#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 #define INT_AMVENCODER INT_DOS_MAILBOX_2
+#define HCODEC_IRQ_MBOX_CLR HCODEC_ASSIST_MBOX2_CLR_REG
+#define HCODEC_IRQ_MBOX_MASK HCODEC_ASSIST_MBOX2_MASK
 #else
-#define AMVENC_DEV_VERSION "AML-MT"
+//#define AMVENC_DEV_VERSION "AML-MT"
 #define INT_AMVENCODER INT_MAILBOX_1A
+#define HCODEC_IRQ_MBOX_CLR HCODEC_ASSIST_MBOX1_CLR_REG
+#define HCODEC_IRQ_MBOX_MASK HCODEC_ASSIST_MBOX1_MASK
 #endif
 
 #define VDEC_166M()  WRITE_MPEG_REG(HHI_VDEC_CLK_CNTL, (5 << 9) | (1 << 8) | (5))
@@ -54,7 +58,6 @@
 #define AMVENC_AVC_IOC_FLUSH_DMA 				_IOW(AMVENC_AVC_IOC_MAGIC, 0x0b, unsigned int)
 #define AMVENC_AVC_IOC_GET_BUFFINFO 				_IOW(AMVENC_AVC_IOC_MAGIC, 0x0c, unsigned int)
 
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 #define AMVENC_AVC_IOC_SET_IE_ME_MB_TYPE 		_IOW(AMVENC_AVC_IOC_MAGIC, 0x0d, unsigned int)
 #define AMVENC_AVC_IOC_SET_ME_PIXEL_MODE 		_IOW(AMVENC_AVC_IOC_MAGIC, 0x0e, unsigned int)
 
@@ -83,7 +86,6 @@ typedef enum{
     FMT_RGBA8888,
     MAX_FRAME_FMT 
 }amvenc_frame_fmt;
-#endif
 
 
 // Memory Address 
@@ -201,7 +203,6 @@ typedef enum{
 #define ANC0_BUFFER_ID					HENC_SCRATCH_A
 #define QPPICTURE						HENC_SCRATCH_B
 
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 #define START_POSITION					HENC_SCRATCH_C
 #define IE_ME_MB_TYPE					HENC_SCRATCH_D
 #define IE_ME_MODE						HENC_SCRATCH_E  //bit 0-4, IE_PIPPELINE_BLOCK, bit 5 me half pixel, bit 6, me step2 sub pixel
@@ -209,7 +210,6 @@ typedef enum{
 
 #define LOW_LATENCY_EN_REG			DOS_SCRATCH9
 #define PREVIOUS_FNUM_REG				DOS_SCRATCH10
-#endif
 
 //---------------------------------------------------
 // ENCODER_STATUS define
diff --git a/drivers/amlogic/amports/m6tvd/ucode/encoder/full_encoder_linux.h b/drivers/amlogic/amports/m6tvd/ucode/encoder/full_encoder_linux.h
index dfb00debe447..fba2c462d0e3 100755
--- a/drivers/amlogic/amports/m6tvd/ucode/encoder/full_encoder_linux.h
+++ b/drivers/amlogic/amports/m6tvd/ucode/encoder/full_encoder_linux.h
@@ -1,178 +1,173 @@
-const u32 MicroCode[] __attribute__((aligned(8))) = {
-	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc013ec0,0x0000000,
-	0xc000e80,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0xc780ec0,0x0000000,
-	0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0xc7928c0,0x0000000,0xc78cc40,0x64d3008,0xc780900,0x0000000,0xc792900,
-	0x0000000,0xcc00000,0x0400000,0x80d1a00,0x80ac300,0x80ac400,0xcc00000,0x0000000,
-	0x64ac808,0x7c00108,0x64ac908,0x7c00148,0x64ac608,0x7c00188,0x64ac508,0x7c001c8,
-	0x64ac708,0x7c00208,0x64aca08,0x7c00388,0x64acb08,0x7c00048,0x64ace08,0x7c003c8,
-	0xcc00000,0x0000000,0xcc00000,0x0000000,0x80d2001,0x64d2008,0xbefc048,0xcc00000,
-	0x0000000,0x6432421,0xaa0c021,0x8032400,0x6699321,0x64ac021,0xaaec021,0xaa34061,
-	0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,0xc784dc0,0x0000000,0xa6100e1,0x0000000,
-	0xc782980,0x0000000,0xc7ffbc0,0x0000000,0xc07f500,0x0000000,0x6bc0008,0x60d1f08,
-	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
-	0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x6a10808,0x60d1f08,
-	0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,0xc07f400,0x80d2100,0x7800148,0x4404208,
-	0xc07f300,0x60d2108,0xc07f280,0x80d2100,0x7800108,0x4404208,0xc07f180,0x60d2108,
-	0xc07f100,0x80d2101,0x80d1f20,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
-	0x7c00308,0xc07eec0,0x60d2108,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,
-	0xc07ed00,0x60d2108,0x80d1f30,0x80d1f30,0x64d3108,0x9210208,0x6810e09,0xa803248,
-	0xc7802c0,0x80d1f20,0x80d1f30,0xc07ea40,0x80d2100,0xc07e9c0,0x80d2100,0xc07e940,
-	0x80d2100,0xc07e8c0,0x80d2104,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,
-	0x80ac007,0x8007801,0xc7fe7c0,0x0000000,0xc07e100,0x0000000,0x6bc0008,0x60d1f08,
-	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
-	0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0xc07e180,0x80d2100,
-	0xc07e100,0x80d2100,0x80d1f20,0x80d1f20,0xc07e000,0x80d2100,0xc07df80,0x80d2100,
-	0xc07df00,0x80d2100,0x80d1f20,0x80d1f40,0x7800048,0xc007d40,0x441a208,0xc07dd40,
-	0x60d2108,0xc007c40,0x6800008,0xc07dc40,0x60d2108,0xc07dbc0,0x80d2100,0x80d1f20,
-	0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,
-	0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,
-	0xbefc108,0x0000000,0x80ac008,0x8007801,0xc7fd340,0x0000000,0xc010400,0x0000000,
-	0x6490b08,0x9c08608,0x9c08648,0x6090b08,0x6401108,0x9c08748,0x9c08788,0x6001108,
-	0x6401008,0x9808008,0x6001008,0x9c08008,0x6001008,0xc010840,0x0000000,0xc00ff40,
-	0x0000000,0xc07c7c0,0x0000000,0x64ac408,0x4001208,0x60ac408,0x64f4908,0x9c08048,
-	0x60f4908,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,
-	0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a19408,0x60d1f08,0x64d1e08,0x98087c8,
-	0x60d1e08,0x64d3308,0x9008248,0x7c00089,0x9108248,0x7c000c9,0x9210208,0xc07c540,
-	0x60d2108,0xc07c4c0,0x80d2107,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
-	0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x6800008,0x7c00248,
-	0x60f1b08,0x7800309,0x946d209,0x60d3208,0x6940008,0x6c00008,0x2008308,0x60f2008,
-	0x6803908,0x64d3309,0x9008249,0x9508209,0x9808508,0x9808548,0x98087c8,0x60f1f08,
-	0x7800309,0x7800348,0x960f248,0x60f1e09,0xc07bc00,0x80d2100,0x7800148,0x7800189,
-	0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x78001c8,0xc07b980,0x60d2108,
-	0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,
-	0x80d1f20,0x64f1c08,0x9206208,0x7800049,0xc005580,0x2408248,0xc07b580,0x60d2108,
-	0x80d3d00,0x64acd08,0x60f4208,0xc0055c0,0x0000000,0x680d683,0x6813344,0x80d300f,
-	0x64d0008,0x9808008,0x9808048,0x9808088,0x9c080c8,0x9c08148,0x9c08188,0x9c081c8,
-	0x60d0008,0x64acf08,0x60f4808,0x6800289,0x78003c8,0x9485248,0x60f4009,0x80ac005,
-	0x8007801,0xc7fb000,0x0000000,0xc00eac0,0x0000000,0x64f4008,0x9c08048,0x60f4008,
-	0xc00df80,0x0000000,0x6490b08,0x9c08608,0x9808648,0x6090b08,0x6401108,0x9808748,
-	0x9c08788,0x6001108,0x6401008,0x9808008,0x6001008,0x9c08008,0x6001008,0xc00e3c0,
-	0x0000000,0xc00dac0,0x0000000,0xc07a340,0x0000000,0x64f4308,0x9c08048,0x9c080c8,
-	0x60f4308,0x64f1508,0x9c08508,0x9c084c8,0x9c08748,0x9c08448,0x9c08488,0x9c08348,
-	0x9c08308,0x9808248,0x9808008,0x60f1508,0x80d3400,0x9808488,0x9808448,0x9808348,
-	0x9808308,0x98085c8,0x9808608,0x9808648,0x9c08688,0x98086c8,0x9c08248,0x9c08008,
-	0x60f1508,0x64f4908,0x9c08048,0x9c081c8,0x9808008,0x60f4908,0x78003c9,0x5805249,
-	0x96e1209,0x5801249,0x9761209,0x98087c8,0x9808788,0x9808748,0x9808688,0x9808648,
-	0x9808608,0x9808288,0x9c08248,0x98081c8,0x9808088,0x9808048,0x60f4908,0x4801249,
-	0xaa10049,0x0000000,0xc781640,0x0000000,0x6800c08,0x6800809,0x94e8209,0x6800c09,
-	0x9608209,0x6801009,0x9708209,0x60f5608,0x6800008,0x6800009,0x9508209,0x60f5708,
-	0x680c0c8,0x6c00008,0x60f5808,0x68000c8,0x68000c9,0x9484209,0x68000c9,0x9504209,
-	0x6800049,0x9581209,0x6800009,0x95a1209,0x6800049,0x95c5209,0x60f5908,0x6800388,
-	0x6800489,0x9508209,0x6800149,0x9604209,0x68000c9,0x9684209,0x6800109,0x9704209,
-	0x6800089,0x9704209,0x60f5a08,0x6800008,0x6800009,0x9508209,0x6800809,0x9608209,
-	0x6801409,0x9708209,0x60f5b08,0x6801008,0x6800009,0x9508209,0x6801809,0x9608209,
-	0x6803009,0x9708209,0x60f5c08,0x6800608,0x6801409,0x9508209,0x6800809,0x9608209,
-	0x6801809,0x9708209,0x60f5d08,0x68000c8,0x6801809,0x948c209,0x6800249,0x9607209,
-	0x6800009,0x96e1209,0x6800c09,0x9708209,0x60f5e08,0x68000c8,0x6800009,0x9461209,
-	0x6800049,0x9481209,0x6800009,0x94a2209,0x6803009,0x950c209,0x6801c09,0x968c209,
-	0x60f5f08,0xc7807c0,0x0000000,0x6804008,0x68000c9,0x9586209,0x6800089,0x9646209,
-	0x9808608,0x9808648,0x9808688,0x98086c8,0x9808708,0x9808748,0x9808788,0x98087c8,
-	0x60f5508,0x6800c08,0x6800809,0x94e8209,0x6800c09,0x9608209,0x6801009,0x9708209,
-	0x60f5608,0x6800c08,0x6801009,0x9508209,0x60f5708,0x680c0c8,0x6c0c0c8,0x60f5808,
-	0x64acc09,0x60f4a09,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,
-	0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x7800309,0x7800348,0x958c248,
-	0x6499008,0x9708248,0x60f4f09,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,
-	0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a10408,0x60d1f08,
-	0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9210208,0xc0773c0,0x60d2108,0xc077340,
-	0x80d2105,0x6800048,0x7c00248,0x60f1b08,0x7800309,0x946d209,0x60d3208,0x7800348,
-	0x960f248,0x60f1e09,0xc077080,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,
-	0x3409289,0x9605248,0x60d1f09,0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,
-	0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x80d1f20,0x64f1d08,0x9206208,0x7800049,
-	0xc000a80,0x2408248,0xc076a80,0x60d2108,0x80d3d00,0x80d5300,0x64acd08,0x60f4208,
-	0xc001840,0x0000000,0x6810143,0x6813bc4,0x6800008,0x6c00448,0x60d3008,0x64d0008,
-	0x9c08008,0x9c08048,0x9c08088,0x9c08188,0x9c081c8,0x60d0008,0x9808008,0x9808048,
-	0x9808088,0x98080c8,0x9808148,0x9808188,0x98081c8,0x60d0008,0x6800048,0x7800309,
-	0x948c209,0x60f4b08,0x80ac005,0x8007801,0xc7f6340,0x0000000,0x0800000,0x7800008,
-	0xc7f5a40,0x6030408,0x6800009,0xb005248,0x2409209,0x5401208,0xcc00000,0x4401208,
-	0xcc00000,0x5401209,0x6bfffda,0x6c0001a,0x6800009,0x680001b,0x6c0001b,0x400071b,
-	0x680000b,0x680c00e,0x680100f,0x740039a,0x400138e,0x7400389,0x400138e,0x740039b,
-	0x400138e,0x740038b,0x400138e,0xaee004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,
-	0x680000f,0x603510e,0x680400d,0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,
-	0xbef804d,0x0000000,0x40403cf,0xb611e8f,0x680800d,0xc7ffd00,0x200e34e,0x64d330d,
-	0x900834d,0x680000e,0x680100f,0xb8053cd,0x0000000,0x404038e,0xc7fff40,0x40403cf,
-	0x7c0028e,0x44013cf,0x7c002cf,0x60d371a,0x60d391b,0x60d381a,0x60d3a1b,0xcc00000,
-	0x0000000,0x8098002,0x8098000,0x6a0001d,0x6d0001d,0x6a0001e,0x6d0001e,0x6a0001f,
-	0x6d0001f,0x6a00020,0x6d00020,0x690000a,0x680001b,0x6c0001b,0x400071b,0x6800009,
-	0x680c00e,0x680064f,0x740039b,0x400138e,0x7400389,0x400138e,0x740039d,0x400138e,
-	0x740038a,0x400138e,0x740039e,0x400138e,0x740038a,0x400138e,0x740039f,0x400138e,
-	0x740038a,0x400138e,0x74003a0,0x400138e,0x740038a,0x400138e,0xaeb004f,0x44013cf,
-	0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,
-	0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40193cf,0xb611e8f,0x6807d0d,
-	0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680064f,0xb8053cd,0x0000000,
-	0x401938e,0xc7fff40,0x40193cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d391b,0x60d4a1d,
-	0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,0x60d4a1d,0x60d3a1b,0x60d4e1d,0x60d4f1e,
-	0x60d501f,0x60d5120,0xcc00000,0x0000000,0xaa10008,0x64d370e,0xcb80003,0x64d390f,
-	0xc7f3440,0x0000000,0x60d371a,0x64d3510,0x908c250,0xa620009,0x60d391b,0x6bfffc9,
-	0x6c00009,0x60d3809,0x6800009,0x6c00009,0x60d3a09,0x9004250,0xaa1c249,0x80d3601,
-	0x80d3603,0x80d3605,0x80d3606,0xc780200,0x80d360f,0x80d3602,0x80d3603,0x80d3604,
-	0x80d3605,0x80d3606,0x80d360f,0x908c210,0x680c10a,0x7800289,0xa814248,0x2409248,
-	0x4401249,0x5402249,0x680c00a,0x200a289,0x740028e,0x400128a,0x581038e,0x740028e,
-	0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400528a,0x78002c9,0xb40d248,
-	0x7800309,0xb40b248,0x700029a,0x400128a,0x700028e,0x961068e,0x400128a,0x700029b,
-	0x400128a,0x700028f,0xc7f25c0,0x96106cf,0x908c210,0x7800289,0x2409248,0x5402249,
-	0x680c00a,0x200a289,0x64d0009,0x9384249,0xa6f8009,0x0000000,0x64d3609,0xa6ec009,
-	0x0000000,0x64d370e,0x64d390f,0x740028e,0x400128a,0x581038e,0x740028e,0x400128a,
-	0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x690000e,0x6c0000e,0x200e30e,
-	0x780028f,0x54033cf,0x200e3ce,0x603510e,0x680400d,0x603520d,0x6a0c00d,0x603500d,
-	0x908c210,0x7800309,0xb830248,0x4001208,0x920c210,0x7800349,0xb82c248,0x6800008,
-	0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,0x6e00008,0x60d1c08,
-	0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,
-	0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,0xbefc108,0x0000000,
-	0x7800188,0x4001208,0x7c00188,0x7800208,0x4002208,0x7c00208,0x80ac009,0x8007801,
-	0xc7f1040,0x0000000,0x7c00288,0x403f248,0x7c002c9,0x5403208,0x690000e,0x6c0000e,
-	0x200e30e,0x200e20e,0x680400f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,
-	0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,0x643500d,
-	0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,0x680c00a,0x64d4a11,0x64d4b12,0x64d4c13,
-	0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x64d3510,0x908c250,0xa62c009,
-	0x60d391b,0x6a00009,0x6d00009,0x60d4e09,0x60d4f09,0x60d5009,0x60d5109,0x6800009,
-	0x6c00009,0x60d3a09,0x80d5201,0x80d3607,0x80d3601,0x80d3608,0x80d3609,0x80d3604,
-	0x80d3605,0x80d3606,0x80d360f,0x908c210,0xaa10008,0x680bfca,0xc000ec0,0x4401208,
-	0x908c210,0x7800309,0xb82b248,0x0000000,0x64d0009,0x9384249,0xa6f8009,0x0000000,
-	0x64d3609,0xa6ec009,0x0000000,0x920c290,0x7800349,0xb80c24a,0x0000000,0x64d5309,
-	0xaa1c009,0x0000000,0x60d2109,0x80d2001,0x64d2009,0xbefc049,0x0000000,0xc7fe040,
-	0x0000000,0x64d390f,0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,0xc0006c0,0x7800308,
-	0xc000e40,0x6800011,0x6a00014,0x6d00014,0x60d4d14,0xc001740,0x680c00a,0x60d4a1d,
-	0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,0x680c28a,0x401528a,0x908c210,0x4002208,
-	0x78002c9,0xb004248,0xc0009c0,0x908c450,0x680c50a,0xc001340,0x0000000,0xc7ef280,
-	0x0000000,0x7800289,0x240a248,0x540324a,0x540128a,0x2009289,0x680c00a,0x200a289,
-	0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x7400291,0x400128a,0x5810451,
-	0x7400291,0x400128a,0x7400292,0x400128a,0x5810492,0x7400292,0x400128a,0x7400293,
-	0x400128a,0x58104d3,0x7400293,0x400128a,0x7400294,0x400128a,0x5810514,0xcc00000,
-	0x7400294,0x690000e,0x6c0000e,0x200e30e,0x780028f,0x540434f,0x54023cf,0x200f34f,
-	0x200e3ce,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,0x603500d,0x7c00291,0x4018251,
-	0x7c002c9,0x5404251,0x5402211,0x2008248,0x690000e,0x6c0000e,0x200e30e,0x200e20e,
-	0x6803e8f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,0x643530d,0x900c34d,
-	0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,0x643500d,0x580f34d,0xbef804d,
-	0xcc00000,0x0000000,0x40006dc,0x700029c,0x400128a,0x7000289,0x9610709,0x400128a,
-	0x700029d,0x400128a,0x7000289,0x9610749,0x400128a,0x700029e,0x400128a,0x7000289,
-	0x9610789,0x400128a,0x700029f,0x400128a,0x7000289,0x96107c9,0x400128a,0x70002a0,
-	0x400128a,0x7000289,0xcc00000,0x9610809,0x64d1602,0x9162082,0xaa0c002,0xc7fff40,
-	0x0000000,0xc7ed800,0x0000000,0xcb80004,0x64f1f08,0x9361248,0xaa10009,0x9042248,
-	0xc7ed640,0x0000000,0xaa14009,0x680324a,0xaa0c049,0x680348a,0x680390a,0x7800088,
-	0x7800309,0xb806248,0x4001208,0x6800008,0x78000c9,0x4001249,0x7c000c9,0x7c00088,
-	0x950c288,0xa60c008,0x7800309,0x980a50a,0xa403248,0x0000000,0x980a58a,0x78000c8,
-	0xa60c008,0x60f4108,0x980a54a,0x980a7ca,0x60f1f0a,0xc7ecf00,0x0000000,0x64f4b08,
-	0x9384248,0xaa10009,0x64f4a08,0xc7ecd80,0x0000000,0x920c248,0x780034a,0xb408289,
-	0x4001249,0x960c209,0x60f4a08,0x6800048,0x7800309,0x948c209,0x60f4b08,0xc7eca80,
-	0x0000000,0x64ac10c,0x680c008,0x6094008,0x6803009,0x6804008,0xd000009,0x6094108,
-	0x6800008,0x6094008,0xcc00000,0x0000000,0xcc00000,0x0000000,0xc000a00,0x0000000,
-	0x780034a,0x400128a,0x40002ca,0x7800309,0x4001249,0x96102c9,0x609520b,0x54042ca,
-	0x968c2c9,0x609080b,0x681010a,0x6c3454a,0x609070a,0x6490b0a,0x980a00a,0x980a0ca,
-	0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,0x698000a,0x6c0000a,0x200a30a,0x609440a,
-	0x6a0000a,0x6c0000a,0x200a30a,0x609450a,0xcc00000,0x0000000,0xc0005c0,0x0000000,
-	0x8095003,0x8095000,0x6820889,0x6c00209,0xcc00000,0x6095109,0x683ffc9,0x649090a,
-	0xa21004a,0xaa0c009,0xc7fff40,0x4401249,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,
-	0x9c0a00a,0x9c0a0ca,0x6090b0a,0xcc00000,0x0000000,0x683ffca,0x649530b,0x90012cb,
-	0x649bb09,0x920c249,0x20092c9,0xaa10009,0xaa0c00a,0xc7ffe40,0x440128a,0xcc00000,
-	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000
+const u32 MicroCode[] __attribute__((aligned(8))) = {
+	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc013dc0,0x0000000,
+	0xc000e80,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0xc780ec0,0x0000000,
+	0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x0000000,0xc7927c0,0x0000000,0xc78cb40,0x64d3008,0xc780900,0x0000000,0xc792800,
+	0x0000000,0xcc00000,0x0400000,0x80d1a00,0x80ac300,0x80ac400,0xcc00000,0x0000000,
+	0x64ac808,0x7c00108,0x64ac908,0x7c00148,0x64ac608,0x7c00188,0x64ac508,0x7c001c8,
+	0x64ac708,0x7c00208,0x64aca08,0x7c00388,0x64acb08,0x7c00048,0x64ace08,0x7c003c8,
+	0xcc00000,0x0000000,0xcc00000,0x0000000,0x80d2001,0x64d2008,0xbefc048,0xcc00000,
+	0x0000000,0x64ac021,0xaafc021,0xaa34061,0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,
+	0xc784dc0,0x0000000,0xa6100e1,0x0000000,0xc782980,0x0000000,0xc7ffcc0,0x0000000,
+	0xc07f600,0x0000000,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,
+	0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,
+	0x98087c8,0x60d1e08,0x6a10808,0x60d1f08,0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,
+	0xc07f500,0x80d2100,0x7800148,0x4404208,0xc07f400,0x60d2108,0xc07f380,0x80d2100,
+	0x7800108,0x4404208,0xc07f280,0x60d2108,0xc07f200,0x80d2101,0x80d1f20,0x64d3109,
+	0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,0xc07efc0,0x60d2108,0x9210209,
+	0x400f208,0x5804208,0x4401208,0x7c00348,0xc07ee00,0x60d2108,0x80d1f30,0x80d1f30,
+	0x64d3108,0x9210208,0x6810e09,0xa803248,0xc7802c0,0x80d1f20,0x80d1f30,0xc07eb40,
+	0x80d2100,0xc07eac0,0x80d2100,0xc07ea40,0x80d2100,0xc07e9c0,0x80d2104,0x80d1f20,
+	0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x80ac007,0x8007401,0xc7fe8c0,0x0000000,
+	0xc07e200,0x0000000,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,
+	0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,
+	0x98087c8,0x60d1e08,0xc07e280,0x80d2100,0xc07e200,0x80d2100,0x80d1f20,0x80d1f20,
+	0xc07e100,0x80d2100,0xc07e080,0x80d2100,0xc07e000,0x80d2100,0x80d1f20,0x80d1f40,
+	0x7800048,0xc007d40,0x441a208,0xc07de40,0x60d2108,0xc007c40,0x6800008,0xc07dd40,
+	0x60d2108,0xc07dcc0,0x80d2100,0x80d1f20,0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,
+	0x6bc0008,0x60d1f08,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x64d1608,0xbefc108,0x0000000,0x6800008,0x6e00008,0x60d1c08,0x80d1d00,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,
+	0x64d1b08,0x9808648,0x60d1b08,0x64d1608,0xbefc108,0x0000000,0x80ac008,0x8007401,
+	0xc7fd440,0x0000000,0xc010400,0x0000000,0x6490b08,0x9c08608,0x9c08648,0x6090b08,
+	0x6401108,0x9c08748,0x9c08788,0x6001108,0x6401008,0x9808008,0x6001008,0x9c08008,
+	0x6001008,0xc010840,0x0000000,0xc00ff40,0x0000000,0xc07c8c0,0x0000000,0x64ac408,
+	0x4001208,0x60ac408,0x64f4908,0x9c08048,0x60f4908,0x6bc0008,0x60d1f08,0x64d1e08,
+	0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,
+	0x6a19408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9008248,0x7c00089,
+	0x9108248,0x7c000c9,0x9210208,0xc07c640,0x60d2108,0xc07c5c0,0x80d2107,0x64d3109,
+	0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,
+	0x4401208,0x7c00348,0x6800008,0x7c00248,0x60f1b08,0x7800309,0x946d209,0x60d3208,
+	0x6940008,0x6c00008,0x2008308,0x60f2008,0x6803908,0x64d3309,0x9008249,0x9508209,
+	0x9808508,0x9808548,0x98087c8,0x60f1f08,0x7800309,0x7800348,0x960f248,0x60f1e09,
+	0xc07bd00,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,
+	0x60d1f09,0x78001c8,0xc07ba80,0x60d2108,0x7800108,0x7800209,0x680040a,0x240a20a,
+	0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x64f1c08,0x9206208,0x7800049,
+	0xc005580,0x2408248,0xc07b680,0x60d2108,0x80d3d00,0x64acd08,0x60f4208,0xc0055c0,
+	0x0000000,0x680d583,0x6813244,0x80d300f,0x64d0008,0x9808008,0x9808048,0x9808088,
+	0x9c080c8,0x9c08148,0x9c08188,0x9c081c8,0x60d0008,0x64acf08,0x60f4808,0x6800289,
+	0x78003c8,0x9485248,0x60f4009,0x80ac005,0x8007401,0xc7fb100,0x0000000,0xc00eac0,
+	0x0000000,0x64f4008,0x9c08048,0x60f4008,0xc00df80,0x0000000,0x6490b08,0x9c08608,
+	0x9808648,0x6090b08,0x6401108,0x9808748,0x9c08788,0x6001108,0x6401008,0x9808008,
+	0x6001008,0x9c08008,0x6001008,0xc00e3c0,0x0000000,0xc00dac0,0x0000000,0xc07a440,
+	0x0000000,0x64f4308,0x9c08048,0x9c080c8,0x60f4308,0x64f1508,0x9c08508,0x9c084c8,
+	0x9c08748,0x9c08448,0x9c08488,0x9c08348,0x9c08308,0x9808248,0x9808008,0x60f1508,
+	0x80d3400,0x9808488,0x9808448,0x9808348,0x9808308,0x98085c8,0x9808608,0x9808648,
+	0x9c08688,0x98086c8,0x9c08248,0x9c08008,0x60f1508,0x64f4908,0x9c08048,0x9c081c8,
+	0x9808008,0x60f4908,0x78003c9,0x5805249,0x96e1209,0x5801249,0x9761209,0x98087c8,
+	0x9808788,0x9808748,0x9808688,0x9808648,0x9808608,0x9808288,0x9c08248,0x98081c8,
+	0x9808088,0x9808048,0x60f4908,0x4801249,0xaa10049,0x0000000,0xc781640,0x0000000,
+	0x6800c08,0x6800809,0x94e8209,0x6800c09,0x9608209,0x6801009,0x9708209,0x60f5608,
+	0x6800008,0x6800009,0x9508209,0x60f5708,0x680c0c8,0x6c00008,0x60f5808,0x68000c8,
+	0x68000c9,0x9484209,0x68000c9,0x9504209,0x6800049,0x9581209,0x6800009,0x95a1209,
+	0x6800049,0x95c5209,0x60f5908,0x6800388,0x6800489,0x9508209,0x6800149,0x9604209,
+	0x68000c9,0x9684209,0x6800109,0x9704209,0x6800089,0x9704209,0x60f5a08,0x6800008,
+	0x6800009,0x9508209,0x6800809,0x9608209,0x6801409,0x9708209,0x60f5b08,0x6801008,
+	0x6800009,0x9508209,0x6801809,0x9608209,0x6803009,0x9708209,0x60f5c08,0x6800608,
+	0x6801409,0x9508209,0x6800809,0x9608209,0x6801809,0x9708209,0x60f5d08,0x68000c8,
+	0x6801809,0x948c209,0x6800249,0x9607209,0x6800009,0x96e1209,0x6800c09,0x9708209,
+	0x60f5e08,0x68000c8,0x6800009,0x9461209,0x6800049,0x9481209,0x6800009,0x94a2209,
+	0x6803009,0x950c209,0x6801c09,0x968c209,0x60f5f08,0xc7807c0,0x0000000,0x6804008,
+	0x68000c9,0x9586209,0x6800089,0x9646209,0x9808608,0x9808648,0x9808688,0x98086c8,
+	0x9808708,0x9808748,0x9808788,0x98087c8,0x60f5508,0x6800c08,0x6800809,0x94e8209,
+	0x6800c09,0x9608209,0x6801009,0x9708209,0x60f5608,0x6800c08,0x6801009,0x9508209,
+	0x60f5708,0x680c0c8,0x6c0c0c8,0x60f5808,0x64acc09,0x60f4a09,0x64d3109,0x9010209,
+	0x400f208,0x5804208,0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,
+	0x7c00348,0x7800309,0x7800348,0x958c248,0x6499008,0x9708248,0x60f4f09,0x6bc0008,
+	0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,
+	0x6c00408,0x60d1f08,0x6a10408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,
+	0x9210208,0xc0774c0,0x60d2108,0xc077440,0x80d2105,0x6800048,0x7c00248,0x60f1b08,
+	0x7800309,0x946d209,0x60d3208,0x7800348,0x960f248,0x60f1e09,0xc077180,0x80d2100,
+	0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x7800108,
+	0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,
+	0x80d1f20,0x64f1d08,0x9206208,0x7800049,0xc000a80,0x2408248,0xc076b80,0x60d2108,
+	0x80d3d00,0x80d5300,0x64acd08,0x60f4208,0xc001840,0x0000000,0x6810043,0x6813ac4,
+	0x6800008,0x6c00448,0x60d3008,0x64d0008,0x9c08008,0x9c08048,0x9c08088,0x9c08188,
+	0x9c081c8,0x60d0008,0x9808008,0x9808048,0x9808088,0x98080c8,0x9808148,0x9808188,
+	0x98081c8,0x60d0008,0x6800048,0x7800309,0x948c209,0x60f4b08,0x80ac005,0x8007401,
+	0xc7f6440,0x0000000,0x0800000,0x7800008,0xc7f5b40,0x6030408,0x6800009,0xb005248,
+	0x2409209,0x5401208,0xcc00000,0x4401208,0xcc00000,0x5401209,0x6bfffda,0x6c0001a,
+	0x6800009,0x680001b,0x6c0001b,0x400071b,0x680000b,0x680c00e,0x680100f,0x740039a,
+	0x400138e,0x7400389,0x400138e,0x740039b,0x400138e,0x740038b,0x400138e,0xaee004f,
+	0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x680400d,0x603520d,
+	0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40403cf,0xb611e8f,
+	0x680800d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680100f,0xb8053cd,
+	0x0000000,0x404038e,0xc7fff40,0x40403cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d371a,
+	0x60d391b,0x60d381a,0x60d3a1b,0xcc00000,0x0000000,0x8098002,0x8098000,0x6a0001d,
+	0x6d0001d,0x6a0001e,0x6d0001e,0x6a0001f,0x6d0001f,0x6a00020,0x6d00020,0x690000a,
+	0x680001b,0x6c0001b,0x400071b,0x6800009,0x680c00e,0x680064f,0x740039b,0x400138e,
+	0x7400389,0x400138e,0x740039d,0x400138e,0x740038a,0x400138e,0x740039e,0x400138e,
+	0x740038a,0x400138e,0x740039f,0x400138e,0x740038a,0x400138e,0x74003a0,0x400138e,
+	0x740038a,0x400138e,0xaeb004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,
+	0x603510e,0x6803e8d,0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,
+	0x0000000,0x40193cf,0xb611e8f,0x6807d0d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,
+	0x680000e,0x680064f,0xb8053cd,0x0000000,0x401938e,0xc7fff40,0x40193cf,0x7c0028e,
+	0x44013cf,0x7c002cf,0x60d391b,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,
+	0x60d4a1d,0x60d3a1b,0x60d4e1d,0x60d4f1e,0x60d501f,0x60d5120,0xcc00000,0x0000000,
+	0xaa10008,0x64d370e,0xcb80003,0x64d390f,0xc7f3540,0x0000000,0x60d371a,0x64d3510,
+	0x908c250,0xa620009,0x60d391b,0x6bfffc9,0x6c00009,0x60d3809,0x6800009,0x6c00009,
+	0x60d3a09,0x9004250,0xaa1c249,0x80d3601,0x80d3603,0x80d3605,0x80d3606,0xc780200,
+	0x80d360f,0x80d3602,0x80d3603,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,
+	0x680c10a,0x7800289,0xa814248,0x2409248,0x4401249,0x5402249,0x680c00a,0x200a289,
+	0x740028e,0x400128a,0x581038e,0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,
+	0x740028f,0x400528a,0x78002c9,0xb40d248,0x7800309,0xb40b248,0x700029a,0x400128a,
+	0x700028e,0x961068e,0x400128a,0x700029b,0x400128a,0x700028f,0xc7f26c0,0x96106cf,
+	0x908c210,0x7800289,0x2409248,0x5402249,0x680c00a,0x200a289,0x64d0009,0x9384249,
+	0xa6f8009,0x0000000,0x64d3609,0xa6ec009,0x0000000,0x64d370e,0x64d390f,0x740028e,
+	0x400128a,0x581038e,0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,
+	0x400128a,0x690000e,0x6c0000e,0x200e30e,0x780028f,0x54033cf,0x200e3ce,0x603510e,
+	0x680400d,0x603520d,0x6a0c00d,0x603500d,0x908c210,0x7800309,0xb830248,0x4001208,
+	0x920c210,0x7800349,0xb82c248,0x6800008,0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,
+	0x0000000,0x6800008,0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,
+	0x60d1b08,0x64d1608,0xbefc108,0x0000000,0x7800188,0x4001208,0x7c00188,0x7800208,
+	0x4002208,0x7c00208,0x80ac009,0x8007401,0xc7f1140,0x0000000,0x7c00288,0x403f248,
+	0x7c002c9,0x5403208,0x690000e,0x6c0000e,0x200e30e,0x200e20e,0x680400f,0x6b0c010,
+	0x643500d,0x580f34d,0xbef804d,0x0000000,0x643530d,0x900c34d,0xa6f800d,0x0000000,
+	0x603510e,0x603520f,0x6035010,0x643500d,0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,
+	0x680c00a,0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,
+	0x60d4d20,0x64d3510,0x908c250,0xa62c009,0x60d391b,0x6a00009,0x6d00009,0x60d4e09,
+	0x60d4f09,0x60d5009,0x60d5109,0x6800009,0x6c00009,0x60d3a09,0x80d5201,0x80d3607,
+	0x80d3601,0x80d3608,0x80d3609,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,
+	0xaa10008,0x680bfca,0xc000ec0,0x4401208,0x908c210,0x7800309,0xb82b248,0x0000000,
+	0x64d0009,0x9384249,0xa6f8009,0x0000000,0x64d3609,0xa6ec009,0x0000000,0x920c290,
+	0x7800349,0xb80c24a,0x0000000,0x64d5309,0xaa1c009,0x0000000,0x60d2109,0x80d2001,
+	0x64d2009,0xbefc049,0x0000000,0xc7fe040,0x0000000,0x64d390f,0x64d4a11,0x64d4b12,
+	0x64d4c13,0x64d4d14,0xc0006c0,0x7800308,0xc000e40,0x6800011,0x6a00014,0x6d00014,
+	0x60d4d14,0xc001740,0x680c00a,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,
+	0x680c28a,0x401528a,0x908c210,0x4002208,0x78002c9,0xb004248,0xc0009c0,0x908c450,
+	0x680c50a,0xc001340,0x0000000,0xc7ef380,0x0000000,0x7800289,0x240a248,0x540324a,
+	0x540128a,0x2009289,0x680c00a,0x200a289,0x740028f,0x400128a,0x58103cf,0x740028f,
+	0x400128a,0x7400291,0x400128a,0x5810451,0x7400291,0x400128a,0x7400292,0x400128a,
+	0x5810492,0x7400292,0x400128a,0x7400293,0x400128a,0x58104d3,0x7400293,0x400128a,
+	0x7400294,0x400128a,0x5810514,0xcc00000,0x7400294,0x690000e,0x6c0000e,0x200e30e,
+	0x780028f,0x540434f,0x54023cf,0x200f34f,0x200e3ce,0x603510e,0x6803e8d,0x603520d,
+	0x6a0c00d,0x603500d,0x7c00291,0x4018251,0x7c002c9,0x5404251,0x5402211,0x2008248,
+	0x690000e,0x6c0000e,0x200e30e,0x200e20e,0x6803e8f,0x6b0c010,0x643500d,0x580f34d,
+	0xbef804d,0x0000000,0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,
+	0x6035010,0x643500d,0x580f34d,0xbef804d,0xcc00000,0x0000000,0x40006dc,0x700029c,
+	0x400128a,0x7000289,0x9610709,0x400128a,0x700029d,0x400128a,0x7000289,0x9610749,
+	0x400128a,0x700029e,0x400128a,0x7000289,0x9610789,0x400128a,0x700029f,0x400128a,
+	0x7000289,0x96107c9,0x400128a,0x70002a0,0x400128a,0x7000289,0xcc00000,0x9610809,
+	0x64d1602,0x9162082,0xaa0c002,0xc7fff40,0x0000000,0xc7ed900,0x0000000,0xcb80004,
+	0x64f1f08,0x9361248,0xaa10009,0x9042248,0xc7ed740,0x0000000,0xaa14009,0x680324a,
+	0xaa0c049,0x680348a,0x680390a,0x7800088,0x7800309,0xb806248,0x4001208,0x6800008,
+	0x78000c9,0x4001249,0x7c000c9,0x7c00088,0x950c288,0xa60c008,0x7800309,0x980a50a,
+	0xa403248,0x0000000,0x980a58a,0x78000c8,0xa60c008,0x60f4108,0x980a54a,0x980a7ca,
+	0x60f1f0a,0xc7ed000,0x0000000,0x64f4b08,0x9384248,0xaa10009,0x64f4a08,0xc7ece80,
+	0x0000000,0x920c248,0x780034a,0xb408289,0x4001249,0x960c209,0x60f4a08,0x6800048,
+	0x7800309,0x948c209,0x60f4b08,0xc7ecb80,0x0000000,0x64ac10c,0x680c008,0x6094008,
+	0x6803009,0x6804008,0xd000009,0x6094108,0x6800008,0x6094008,0xcc00000,0x0000000,
+	0xcc00000,0x0000000,0xc000a00,0x0000000,0x780034a,0x400128a,0x40002ca,0x7800309,
+	0x4001249,0x96102c9,0x609520b,0x54042ca,0x968c2c9,0x609080b,0x681010a,0x6c3454a,
+	0x609070a,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,
+	0x698000a,0x6c0000a,0x200a30a,0x609440a,0x6a0000a,0x6c0000a,0x200a30a,0x609450a,
+	0xcc00000,0x0000000,0xc0005c0,0x0000000,0x8095003,0x8095000,0x6820889,0x6c00209,
+	0xcc00000,0x6095109,0x683ffc9,0x649090a,0xa21004a,0xaa0c009,0xc7fff40,0x4401249,
+	0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,0xcc00000,
+	0x0000000,0x683ffca,0x649530b,0x90012cb,0x649bb09,0x920c249,0x20092c9,0xaa10009,
+	0xaa0c00a,0xc7ffe40,0x440128a,0xcc00000,0x0000000,0x0000000,0x0000000,0x0000000,
 };
 
-
-
-
-                   
diff --git a/drivers/amlogic/amports/m6tvd/ucode/encoder/half_encoder_linux.h b/drivers/amlogic/amports/m6tvd/ucode/encoder/half_encoder_linux.h
index bfa820952e56..fbea40d7986e 100755
--- a/drivers/amlogic/amports/m6tvd/ucode/encoder/half_encoder_linux.h
+++ b/drivers/amlogic/amports/m6tvd/ucode/encoder/half_encoder_linux.h
@@ -1,144 +1,144 @@
 const u32 MicroCode[] __attribute__((aligned(8))) ={
-	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc010200,0x0000000,
-	0xc000e00,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0x69fffe2,0x6c01c22,
-	0xc780dc0,0x0000000,0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc010080,0x0000000,
+	0xc000e00,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0xc780e40,0x0000000,
+	0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0xc78f100,0x0000000,0xc789480,0x64d3008,0xc780880,0x0000000,0xc78f140,
+	0x0000000,0xc78ef80,0x0000000,0xc789300,0x64d3008,0xc780880,0x0000000,0xc78efc0,
 	0x64f1f08,0xcc00000,0x0400000,0x80d1a00,0x80ac300,0x80ac400,0xcc00000,0x0000000,
 	0x64ac808,0x7c00108,0x64ac908,0x7c00148,0x64ac608,0x7c00188,0x64ac508,0x7c001c8,
 	0x64ac708,0x7c00208,0x64aca08,0x7c00388,0x64acb08,0x7c00048,0xcc00000,0x0000000,
-	0xcc00000,0x0000000,0x80d2001,0x64d2008,0xbefc048,0xcc00000,0x0000000,0xa614022,
-	0x0000000,0x6696022,0x69fffe2,0x6c01c22,0x44018a2,0x64ac021,0xaae4021,0xaa34061,
-	0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,0xc784780,0x0000000,0xa6100e1,0x0000000,
-	0xc782980,0x0000000,0xc7ffb40,0x0000000,0xc07f500,0x0000000,0x6bc0008,0x60d1f08,
-	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
-	0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x6a10808,0x60d1f08,
-	0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,0xc07f380,0x80d2100,0x7800148,0x4404208,
-	0xc07f280,0x60d2108,0xc07f200,0x80d2100,0x7800108,0x4404208,0xc07f100,0x60d2108,
-	0xc07f080,0x80d2101,0x80d1f20,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
-	0x7c00308,0xc07ee40,0x60d2108,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,
-	0xc07ec80,0x60d2108,0x80d1f30,0x80d1f30,0x64d3108,0x9210208,0x6810e09,0xa803248,
-	0xc7802c0,0x80d1f20,0x80d1f30,0xc07e9c0,0x80d2100,0xc07e940,0x80d2100,0xc07e8c0,
-	0x80d2100,0xc07e840,0x80d2104,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,
-	0x80ac007,0x8007801,0xc7fe740,0x0000000,0xc07e100,0x0000000,0x6bc0008,0x60d1f08,
-	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
-	0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0xc07e100,0x80d2100,
-	0xc07e080,0x80d2100,0x80d1f20,0x80d1f20,0xc07df80,0x80d2100,0xc07df00,0x80d2100,
-	0xc07de80,0x80d2100,0x80d1f20,0x80d1f40,0x7800048,0xc004600,0x441a208,0xc07dcc0,
-	0x60d2108,0xc004500,0x6800008,0xc07dbc0,0x60d2108,0xc07db40,0x80d2100,0x80d1f20,
-	0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,
-	0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,
-	0xbefc108,0x0000000,0x80ac008,0x8007801,0xc7fd2c0,0x0000000,0xc00c740,0x0000000,
-	0x6490b08,0x9c08608,0x9c08648,0x6090b08,0xc00cdc0,0x0000000,0xc00c4c0,0x0000000,
-	0xc07ca00,0x0000000,0x64ac408,0x4001208,0x60ac408,0x6bc0008,0x60d1f08,0x64d1e08,
-	0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,
-	0x6a19408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9008248,0x7c00089,
-	0x9108248,0x7c000c9,0x9210208,0xc07c7c0,0x60d2108,0xc07c740,0x80d2107,0x64d3109,
-	0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,
-	0x4401208,0x7c00348,0x6800008,0x7c00248,0x60f1b08,0x7800309,0x946d209,0x60d3208,
-	0x6940008,0x6c00008,0x2008308,0x60f2008,0x6803908,0x64d3309,0x9008249,0x9508209,
-	0x9808508,0x9808548,0x98087c8,0x60f1f08,0x7800309,0x7800348,0x960f248,0x60f1e09,
-	0xc07be80,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,
-	0x60d1f09,0x78001c8,0xc07bc00,0x60d2108,0x7800108,0x7800209,0x680040a,0x240a20a,
-	0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x64f1c08,0x9206208,0x7800049,
-	0xc002140,0x2408248,0xc07b800,0x60d2108,0x80d3d00,0xc002200,0x0000000,0x6809ec3,
-	0x80d300f,0x64d0008,0x9808008,0x9808048,0x9808088,0x60d0008,0x80ac005,0x8007801,
-	0xc7fb5c0,0x0000000,0xc00aa40,0x0000000,0x6490b08,0x9808608,0x9c08648,0x6090b08,
-	0xc00b0c0,0x0000000,0xc00a7c0,0x0000000,0xc07ad00,0x0000000,0x64f1508,0x9c08508,
-	0x9c084c8,0x9c08588,0x9c08448,0x9c08488,0x60f1508,0x9808488,0x9808448,0x98085c8,
-	0x9808608,0x98086c8,0x60f1508,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
-	0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x6bc0008,0x60d1f08,
-	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
-	0x60d1f08,0x6a10408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9210208,
-	0xc07a680,0x60d2108,0xc07a600,0x80d2105,0x6800048,0x7c00248,0x60f1b08,0x7800309,
-	0x946d209,0x60d3208,0x7800309,0x7800348,0x960f248,0x60f1e09,0xc07a300,0x80d2100,
-	0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x7800108,
-	0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,
-	0x80d1f20,0x64f1d08,0x9206208,0x7800049,0xc000640,0x2408248,0xc079d00,0x60d2108,
-	0x80d3d00,0x80d5300,0xc001480,0x0000000,0x680c983,0x6800008,0x6c00448,0x60d3008,
-	0x64d0008,0x9808008,0x9808048,0x9808088,0x60d0008,0x80ac005,0x8007801,0xc7f9a00,
-	0x0000000,0x0800000,0x7800008,0xc7f9180,0x6030408,0x6800009,0xb005248,0x2409209,
-	0x5401208,0xcc00000,0x4401208,0xcc00000,0x5401209,0x6bfffda,0x6c0001a,0x6800009,
-	0x680001b,0x6c0001b,0x400071b,0x680000b,0x680c00e,0x680100f,0x740039a,0x400138e,
-	0x7400389,0x400138e,0x740039b,0x400138e,0x740038b,0x400138e,0xaee004f,0x44013cf,
-	0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x680400d,0x603520d,0x6a0c00d,
-	0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40403cf,0xb611e8f,0x680800d,
-	0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680100f,0xb8053cd,0x0000000,
-	0x404038e,0xc7fff40,0x40403cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d371a,0x60d391b,
-	0x60d381a,0x60d3a1b,0xcc00000,0x0000000,0x6a0001d,0x6d0001d,0x6a0001e,0x6d0001e,
-	0x6a0001f,0x6d0001f,0x6a00020,0x6d00020,0x690000a,0x680001b,0x6c0001b,0x400071b,
-	0x6800009,0x680c00e,0x680064f,0x740039b,0x400138e,0x7400389,0x400138e,0x740039d,
-	0x400138e,0x740038a,0x400138e,0x740039e,0x400138e,0x740038a,0x400138e,0x740039f,
-	0x400138e,0x740038a,0x400138e,0x74003a0,0x400138e,0x740038a,0x400138e,0xaeb004f,
-	0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x6803e8d,0x603520d,
-	0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40193cf,0xb611e8f,
-	0x6807d0d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680064f,0xb8053cd,
-	0x0000000,0x401938e,0xc7fff40,0x40193cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d391b,
-	0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,0x60d4a1d,0x60d3a1b,0x60d4e1d,
-	0x60d4f1e,0x60d501f,0x60d5120,0xcc00000,0x0000000,0xaa10008,0x64d370e,0xcb80003,
-	0x64d390f,0xc7f6c00,0x0000000,0x60d371a,0x64d3510,0x908c250,0xa620009,0x60d391b,
-	0x6bfffc9,0x6c00009,0x60d3809,0x6800009,0x6c00009,0x60d3a09,0x9004250,0xaa1c249,
-	0x80d3601,0x80d3603,0x80d3605,0x80d3606,0xc780200,0x80d360f,0x80d3602,0x80d3603,
-	0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0x680c10a,0x7800289,0xa814248,
-	0x2409248,0x4401249,0x5402249,0x680c00a,0x200a289,0x740028e,0x400128a,0x581038e,
-	0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400528a,0x78002c9,
-	0xb40d248,0x7800309,0xb40b248,0x700029a,0x400128a,0x700028e,0x961068e,0x400128a,
-	0x700029b,0x400128a,0x700028f,0xc7f5d80,0x96106cf,0x908c210,0x7800289,0x2409248,
-	0x5402249,0x680c00a,0x200a289,0x64d0009,0x9384249,0xa6f8009,0x0000000,0x64d3609,
-	0xa6ec009,0x0000000,0x64d370e,0x64d390f,0x740028e,0x400128a,0x581038e,0x740028e,
-	0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x690000e,0x6c0000e,
-	0x200e30e,0x780028f,0x54033cf,0x200e3ce,0x603510e,0x680400d,0x603520d,0x6a0c00d,
-	0x603500d,0x908c210,0x7800309,0xb830248,0x4001208,0x920c210,0x7800349,0xb82c248,
-	0x6800008,0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,0x6e00008,
-	0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,0xbefc108,
-	0x0000000,0x7800188,0x4001208,0x7c00188,0x7800208,0x4002208,0x7c00208,0x80ac009,
-	0x8007801,0xc7f4800,0x0000000,0x7c00288,0x403f248,0x7c002c9,0x5403208,0x690000e,
-	0x6c0000e,0x200e30e,0x200e20e,0x680400f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,
+	0xcc00000,0x0000000,0x80d2001,0x64d2008,0xbefc048,0xcc00000,0x0000000,0x64ac021,
+	0xaafc021,0xaa34061,0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,0xc784780,0x0000000,
+	0xa6100e1,0x0000000,0xc782980,0x0000000,0xc7ffcc0,0x0000000,0xc07f680,0x0000000,
+	0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,
+	0x6800048,0x6c00408,0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,
+	0x6a10808,0x60d1f08,0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,0xc07f500,0x80d2100,
+	0x7800148,0x4404208,0xc07f400,0x60d2108,0xc07f380,0x80d2100,0x7800108,0x4404208,
+	0xc07f280,0x60d2108,0xc07f200,0x80d2101,0x80d1f20,0x64d3109,0x9010209,0x400f208,
+	0x5804208,0x4401208,0x7c00308,0xc07efc0,0x60d2108,0x9210209,0x400f208,0x5804208,
+	0x4401208,0x7c00348,0xc07ee00,0x60d2108,0x80d1f30,0x80d1f30,0x64d3108,0x9210208,
+	0x6810e09,0xa803248,0xc7802c0,0x80d1f20,0x80d1f30,0xc07eb40,0x80d2100,0xc07eac0,
+	0x80d2100,0xc07ea40,0x80d2100,0xc07e9c0,0x80d2104,0x80d1f20,0x80d1f30,0x6bc0008,
+	0x60d1f08,0x0000000,0x80ac007,0x8007401,0xc7fe8c0,0x0000000,0xc07e280,0x0000000,
+	0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,
+	0x6800048,0x6c00408,0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,
+	0xc07e280,0x80d2100,0xc07e200,0x80d2100,0x80d1f20,0x80d1f20,0xc07e100,0x80d2100,
+	0xc07e080,0x80d2100,0xc07e000,0x80d2100,0x80d1f20,0x80d1f40,0x7800048,0xc004600,
+	0x441a208,0xc07de40,0x60d2108,0xc004500,0x6800008,0xc07dd40,0x60d2108,0xc07dcc0,
+	0x80d2100,0x80d1f20,0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,
+	0x0000000,0x6800008,0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,
+	0x60d1b08,0x64d1608,0xbefc108,0x0000000,0x80ac008,0x8007401,0xc7fd440,0x0000000,
+	0xc00c740,0x0000000,0x6490b08,0x9c08608,0x9c08648,0x6090b08,0xc00cdc0,0x0000000,
+	0xc00c4c0,0x0000000,0xc07cb80,0x0000000,0x64ac408,0x4001208,0x60ac408,0x6bc0008,
+	0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,
+	0x6c00408,0x60d1f08,0x6a19408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,
+	0x9008248,0x7c00089,0x9108248,0x7c000c9,0x9210208,0xc07c940,0x60d2108,0xc07c8c0,
+	0x80d2107,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,0x9210209,
+	0x400f208,0x5804208,0x4401208,0x7c00348,0x6800008,0x7c00248,0x60f1b08,0x7800309,
+	0x946d209,0x60d3208,0x6940008,0x6c00008,0x2008308,0x60f2008,0x6803908,0x64d3309,
+	0x9008249,0x9508209,0x9808508,0x9808548,0x98087c8,0x60f1f08,0x7800309,0x7800348,
+	0x960f248,0x60f1e09,0xc07c000,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,
+	0x3409289,0x9605248,0x60d1f09,0x78001c8,0xc07bd80,0x60d2108,0x7800108,0x7800209,
+	0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x64f1c08,
+	0x9206208,0x7800049,0xc002140,0x2408248,0xc07b980,0x60d2108,0x80d3d00,0xc002200,
+	0x0000000,0x6809d43,0x80d300f,0x64d0008,0x9808008,0x9808048,0x9808088,0x60d0008,
+	0x80ac005,0x8007401,0xc7fb740,0x0000000,0xc00aa40,0x0000000,0x6490b08,0x9808608,
+	0x9c08648,0x6090b08,0xc00b0c0,0x0000000,0xc00a7c0,0x0000000,0xc07ae80,0x0000000,
+	0x64f1508,0x9c08508,0x9c084c8,0x9c08588,0x9c08448,0x9c08488,0x60f1508,0x9808488,
+	0x9808448,0x98085c8,0x9808608,0x98086c8,0x60f1508,0x64d3109,0x9010209,0x400f208,
+	0x5804208,0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,
+	0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,
+	0x6800048,0x6c00408,0x60d1f08,0x6a10408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,
+	0x64d3308,0x9210208,0xc07a800,0x60d2108,0xc07a780,0x80d2105,0x6800048,0x7c00248,
+	0x60f1b08,0x7800309,0x946d209,0x60d3208,0x7800309,0x7800348,0x960f248,0x60f1e09,
+	0xc07a480,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,
+	0x60d1f09,0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,
+	0x80d1f20,0x80d1f20,0x80d1f20,0x64f1d08,0x9206208,0x7800049,0xc000640,0x2408248,
+	0xc079e80,0x60d2108,0x80d3d00,0x80d5300,0xc001480,0x0000000,0x680c803,0x6800008,
+	0x6c00448,0x60d3008,0x64d0008,0x9808008,0x9808048,0x9808088,0x60d0008,0x80ac005,
+	0x8007401,0xc7f9b80,0x0000000,0x0800000,0x7800008,0xc7f9300,0x6030408,0x6800009,
+	0xb005248,0x2409209,0x5401208,0xcc00000,0x4401208,0xcc00000,0x5401209,0x6bfffda,
+	0x6c0001a,0x6800009,0x680001b,0x6c0001b,0x400071b,0x680000b,0x680c00e,0x680100f,
+	0x740039a,0x400138e,0x7400389,0x400138e,0x740039b,0x400138e,0x740038b,0x400138e,
+	0xaee004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x680400d,
+	0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40403cf,
+	0xb611e8f,0x680800d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680100f,
+	0xb8053cd,0x0000000,0x404038e,0xc7fff40,0x40403cf,0x7c0028e,0x44013cf,0x7c002cf,
+	0x60d371a,0x60d391b,0x60d381a,0x60d3a1b,0xcc00000,0x0000000,0x6a0001d,0x6d0001d,
+	0x6a0001e,0x6d0001e,0x6a0001f,0x6d0001f,0x6a00020,0x6d00020,0x690000a,0x680001b,
+	0x6c0001b,0x400071b,0x6800009,0x680c00e,0x680064f,0x740039b,0x400138e,0x7400389,
+	0x400138e,0x740039d,0x400138e,0x740038a,0x400138e,0x740039e,0x400138e,0x740038a,
+	0x400138e,0x740039f,0x400138e,0x740038a,0x400138e,0x74003a0,0x400138e,0x740038a,
+	0x400138e,0xaeb004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,
+	0x6803e8d,0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,
+	0x40193cf,0xb611e8f,0x6807d0d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,
+	0x680064f,0xb8053cd,0x0000000,0x401938e,0xc7fff40,0x40193cf,0x7c0028e,0x44013cf,
+	0x7c002cf,0x60d391b,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,0x60d4a1d,
+	0x60d3a1b,0x60d4e1d,0x60d4f1e,0x60d501f,0x60d5120,0xcc00000,0x0000000,0xaa10008,
+	0x64d370e,0xcb80003,0x64d390f,0xc7f6d80,0x0000000,0x60d371a,0x64d3510,0x908c250,
+	0xa620009,0x60d391b,0x6bfffc9,0x6c00009,0x60d3809,0x6800009,0x6c00009,0x60d3a09,
+	0x9004250,0xaa1c249,0x80d3601,0x80d3603,0x80d3605,0x80d3606,0xc780200,0x80d360f,
+	0x80d3602,0x80d3603,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0x680c10a,
+	0x7800289,0xa814248,0x2409248,0x4401249,0x5402249,0x680c00a,0x200a289,0x740028e,
+	0x400128a,0x581038e,0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,
+	0x400528a,0x78002c9,0xb40d248,0x7800309,0xb40b248,0x700029a,0x400128a,0x700028e,
+	0x961068e,0x400128a,0x700029b,0x400128a,0x700028f,0xc7f5f00,0x96106cf,0x908c210,
+	0x7800289,0x2409248,0x5402249,0x680c00a,0x200a289,0x64d0009,0x9384249,0xa6f8009,
+	0x0000000,0x64d3609,0xa6ec009,0x0000000,0x64d370e,0x64d390f,0x740028e,0x400128a,
+	0x581038e,0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,
+	0x690000e,0x6c0000e,0x200e30e,0x780028f,0x54033cf,0x200e3ce,0x603510e,0x680400d,
+	0x603520d,0x6a0c00d,0x603500d,0x908c210,0x7800309,0xb830248,0x4001208,0x920c210,
+	0x7800349,0xb82c248,0x6800008,0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,
+	0x6800008,0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,
+	0x64d1608,0xbefc108,0x0000000,0x7800188,0x4001208,0x7c00188,0x7800208,0x4002208,
+	0x7c00208,0x80ac009,0x8007401,0xc7f4980,0x0000000,0x7c00288,0x403f248,0x7c002c9,
+	0x5403208,0x690000e,0x6c0000e,0x200e30e,0x200e20e,0x680400f,0x6b0c010,0x643500d,
+	0x580f34d,0xbef804d,0x0000000,0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,
+	0x603520f,0x6035010,0x643500d,0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,0x680c00a,
+	0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,
+	0x64d3510,0x908c250,0xa62c009,0x60d391b,0x6a00009,0x6d00009,0x60d4e09,0x60d4f09,
+	0x60d5009,0x60d5109,0x6800009,0x6c00009,0x60d3a09,0x80d5201,0x80d3607,0x80d3601,
+	0x80d3608,0x80d3609,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0xaa10008,
+	0x680bfca,0xc000ec0,0x4401208,0x908c210,0x7800309,0xb82b248,0x0000000,0x64d0009,
+	0x9384249,0xa6f8009,0x0000000,0x64d3609,0xa6ec009,0x0000000,0x920c290,0x7800349,
+	0xb80c24a,0x0000000,0x64d5309,0xaa1c009,0x0000000,0x60d2109,0x80d2001,0x64d2009,
+	0xbefc049,0x0000000,0xc7fe040,0x0000000,0x64d390f,0x64d4a11,0x64d4b12,0x64d4c13,
+	0x64d4d14,0xc0006c0,0x7800308,0xc000e40,0x6800011,0x6a00014,0x6d00014,0x60d4d14,
+	0xc001740,0x680c00a,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,0x680c28a,
+	0x401528a,0x908c210,0x4002208,0x78002c9,0xb004248,0xc0009c0,0x908c450,0x680c50a,
+	0xc001340,0x0000000,0xc7f2bc0,0x0000000,0x7800289,0x240a248,0x540324a,0x540128a,
+	0x2009289,0x680c00a,0x200a289,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,
+	0x7400291,0x400128a,0x5810451,0x7400291,0x400128a,0x7400292,0x400128a,0x5810492,
+	0x7400292,0x400128a,0x7400293,0x400128a,0x58104d3,0x7400293,0x400128a,0x7400294,
+	0x400128a,0x5810514,0xcc00000,0x7400294,0x690000e,0x6c0000e,0x200e30e,0x780028f,
+	0x540434f,0x54023cf,0x200f34f,0x200e3ce,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,
+	0x603500d,0x7c00291,0x4018251,0x7c002c9,0x5404251,0x5402211,0x2008248,0x690000e,
+	0x6c0000e,0x200e30e,0x200e20e,0x6803e8f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,
 	0x0000000,0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,
-	0x643500d,0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,0x680c00a,0x64d4a11,0x64d4b12,
-	0x64d4c13,0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x64d3510,0x908c250,
-	0xa62c009,0x60d391b,0x6a00009,0x6d00009,0x60d4e09,0x60d4f09,0x60d5009,0x60d5109,
-	0x6800009,0x6c00009,0x60d3a09,0x80d5201,0x80d3607,0x80d3601,0x80d3608,0x80d3609,
-	0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0xaa10008,0x680bfca,0xc000ec0,
-	0x4401208,0x908c210,0x7800309,0xb82b248,0x0000000,0x64d0009,0x9384249,0xa6f8009,
-	0x0000000,0x64d3609,0xa6ec009,0x0000000,0x920c290,0x7800349,0xb80c24a,0x0000000,
-	0x64d5309,0xaa1c009,0x0000000,0x60d2109,0x80d2001,0x64d2009,0xbefc049,0x0000000,
-	0xc7fe040,0x0000000,0x64d390f,0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,0xc0006c0,
-	0x7800308,0xc000e40,0x6800011,0x6a00014,0x6d00014,0x60d4d14,0xc001740,0x680c00a,
-	0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,0x680c28a,0x401528a,0x908c210,
-	0x4002208,0x78002c9,0xb004248,0xc0009c0,0x908c450,0x680c50a,0xc001340,0x0000000,
-	0xc7f2a40,0x0000000,0x7800289,0x240a248,0x540324a,0x540128a,0x2009289,0x680c00a,
-	0x200a289,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x7400291,0x400128a,
-	0x5810451,0x7400291,0x400128a,0x7400292,0x400128a,0x5810492,0x7400292,0x400128a,
-	0x7400293,0x400128a,0x58104d3,0x7400293,0x400128a,0x7400294,0x400128a,0x5810514,
-	0xcc00000,0x7400294,0x690000e,0x6c0000e,0x200e30e,0x780028f,0x540434f,0x54023cf,
-	0x200f34f,0x200e3ce,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,0x603500d,0x7c00291,
-	0x4018251,0x7c002c9,0x5404251,0x5402211,0x2008248,0x690000e,0x6c0000e,0x200e30e,
-	0x200e20e,0x6803e8f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,0x643530d,
-	0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,0x643500d,0x580f34d,
-	0xbef804d,0xcc00000,0x0000000,0x40006dc,0x700029c,0x400128a,0x7000289,0x9610709,
-	0x400128a,0x700029d,0x400128a,0x7000289,0x9610749,0x400128a,0x700029e,0x400128a,
-	0x7000289,0x9610789,0x400128a,0x700029f,0x400128a,0x7000289,0x96107c9,0x400128a,
-	0x70002a0,0x400128a,0x7000289,0xcc00000,0x9610809,0x64d1602,0x9162082,0xaa0c002,
-	0xc7fff40,0x0000000,0xc7f0fc0,0x0000000,0x93e1248,0xaa10009,0x9042248,0xc7f0e80,
-	0x0000000,0xaa14009,0x680324a,0xaa0c049,0x680348a,0x680390a,0x7800088,0x7800309,
-	0xb806248,0x4001208,0x6800008,0x78000c9,0x4001249,0x7c000c9,0x7c00088,0x950c288,
-	0xa60c008,0x7800309,0x980a50a,0xa403248,0x0000000,0x980a58a,0x78000c8,0xa60c008,
-	0x0000000,0x980a54a,0x980a7ca,0x60f1f0a,0xc7f0740,0x0000000,0x64ac10c,0x680c008,
-	0x6094008,0x6803009,0x6804008,0xd000009,0x6094108,0x6800008,0x6094008,0xcc00000,
-	0x0000000,0xcc00000,0x0000000,0xc000a00,0x0000000,0x780034a,0x400128a,0x40002ca,
-	0x7800309,0x4001249,0x96102c9,0x609520b,0x54042ca,0x968c2c9,0x609080b,0x681010a,
-	0x6c3454a,0x609070a,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,
-	0x6090b0a,0x698000a,0x6c0000a,0x200a30a,0x609440a,0x6a0000a,0x6c0000a,0x200a30a,
-	0x609450a,0xcc00000,0x0000000,0xc0005c0,0x0000000,0x8095003,0x8095000,0x6820889,
-	0x6c00209,0xcc00000,0x6095109,0x683ffc9,0x649090a,0xa21004a,0xaa0c009,0xc7fff40,
-	0x4401249,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,
-	0xcc00000,0x0000000,0x683ffca,0x649530b,0x90012cb,0x649bb09,0x920c249,0x20092c9,
-	0xaa10009,0xaa0c00a,0xc7ffe40,0x440128a,0xcc00000,0x0000000,0x0000000,0x0000000,
+	0x643500d,0x580f34d,0xbef804d,0xcc00000,0x0000000,0x40006dc,0x700029c,0x400128a,
+	0x7000289,0x9610709,0x400128a,0x700029d,0x400128a,0x7000289,0x9610749,0x400128a,
+	0x700029e,0x400128a,0x7000289,0x9610789,0x400128a,0x700029f,0x400128a,0x7000289,
+	0x96107c9,0x400128a,0x70002a0,0x400128a,0x7000289,0xcc00000,0x9610809,0x64d1602,
+	0x9162082,0xaa0c002,0xc7fff40,0x0000000,0xc7f1140,0x0000000,0x93e1248,0xaa10009,
+	0x9042248,0xc7f1000,0x0000000,0xaa14009,0x680324a,0xaa0c049,0x680348a,0x680390a,
+	0x7800088,0x7800309,0xb806248,0x4001208,0x6800008,0x78000c9,0x4001249,0x7c000c9,
+	0x7c00088,0x950c288,0xa60c008,0x7800309,0x980a50a,0xa403248,0x0000000,0x980a58a,
+	0x78000c8,0xa60c008,0x0000000,0x980a54a,0x980a7ca,0x60f1f0a,0xc7f08c0,0x0000000,
+	0x64ac10c,0x680c008,0x6094008,0x6803009,0x6804008,0xd000009,0x6094108,0x6800008,
+	0x6094008,0xcc00000,0x0000000,0xcc00000,0x0000000,0xc000a00,0x0000000,0x780034a,
+	0x400128a,0x40002ca,0x7800309,0x4001249,0x96102c9,0x609520b,0x54042ca,0x968c2c9,
+	0x609080b,0x681010a,0x6c3454a,0x609070a,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,
+	0x9c0a00a,0x9c0a0ca,0x6090b0a,0x698000a,0x6c0000a,0x200a30a,0x609440a,0x6a0000a,
+	0x6c0000a,0x200a30a,0x609450a,0xcc00000,0x0000000,0xc0005c0,0x0000000,0x8095003,
+	0x8095000,0x6820889,0x6c00209,0xcc00000,0x6095109,0x683ffc9,0x649090a,0xa21004a,
+	0xaa0c009,0xc7fff40,0x4401249,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,
+	0x9c0a0ca,0x6090b0a,0xcc00000,0x0000000,0x683ffca,0x649530b,0x90012cb,0x649bb09,
+	0x920c249,0x20092c9,0xaa10009,0xaa0c00a,0xc7ffe40,0x440128a,0xcc00000,0x0000000,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 };
 
 
diff --git a/drivers/amlogic/amports/vdec_reg.h b/drivers/amlogic/amports/vdec_reg.h
index 76db8a7b2c92..dcded3496645 100755
--- a/drivers/amlogic/amports/vdec_reg.h
+++ b/drivers/amlogic/amports/vdec_reg.h
@@ -37,8 +37,8 @@
 #define SET_VREG_MASK(r, mask) WRITE_VREG(r, READ_VREG(r) | (mask))
 #define CLEAR_VREG_MASK(r, mask) WRITE_VREG(r, READ_VREG(r) & ~(mask))
 
-#define READ_HREG(r) (__raw_readl((volatile void __iomem *)DOS_REG_ADDR(r)))
-#define WRITE_HREG(r, val) __raw_writel(val, (volatile void __iomem *)DOS_REG_ADDR(r))
+#define READ_HREG(r) (__raw_readl((volatile void __iomem *)DOS_REG_ADDR(r|0x1000)))
+#define WRITE_HREG(r, val) __raw_writel(val, (volatile void __iomem *)DOS_REG_ADDR(r|0x1000))
 #define WRITE_HREG_BITS(r, val, start, len) \
     WRITE_HREG(r, (READ_HREG(r) & ~(((1L<<(len))-1)<<(start)))|((unsigned)((val)&((1L<<(len))-1)) << (start)))
 #define SET_HREG_MASK(r, mask) WRITE_HREG(r, READ_HREG(r) | (mask))
diff --git a/include/linux/amlogic/amports/canvas.h b/include/linux/amlogic/amports/canvas.h
index 84530289b87a..cfcdda6152bc 100755
--- a/include/linux/amlogic/amports/canvas.h
+++ b/include/linux/amlogic/amports/canvas.h
@@ -149,10 +149,10 @@ typedef struct {
 #define D2D3_CANVAS_DBR_INDEX      0xEE
 #elif MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON6TV
 #define AMVENC_CANVAS_INDEX 0xDB
-#define AMVENC_CANVAS_MAX_INDEX 0xE0
+#define AMVENC_CANVAS_MAX_INDEX 0xE3
 
-#define D2D3_CANVAS_DPG_INDEX      0xE1
-#define D2D3_CANVAS_DBR_INDEX      0xE2
+#define D2D3_CANVAS_DPG_INDEX      0xE4
+#define D2D3_CANVAS_DBR_INDEX      0xE5
 #endif
 
 extern void canvas_config(u32 index, ulong addr, u32 width,
-- 
2.19.0

